src/receive_imf.rs

```diff
@@ -1632,86 +1637,136 @@ async fn apply_group_changes(
         return Ok(None);
     }
 
-    let mut recreate_member_list = false;
     let mut send_event_chat_modified = false;
-
+    let mut removed_id = None;
     let mut better_msg = None;
-    let removed_id;
+
+    let self_added =
+        if let Some(added_addr) = mime_parser.get_header(HeaderDef::ChatGroupMemberAdded) {
+            if let Ok(self_addr) = context.get_primary_self_addr().await {
+                self_addr == *added_addr
+            } else {
+                false
+            }
+        } else {
+            false
+        };
+
+    let recreate_member_list =
+        // if we're not in the group, reject don't bother about group members
+        if !chat::is_contact_in_chat(context, chat_id, ContactId::SELF).await? && !self_added {
+            false
+        } else {
+            let apply_group_changes = chat_id
+                .update_timestamp(context, Param::MemberListTimestamp, sent_timestamp)
+                .await?;
+
+            // always reject old group changes
+
+            if apply_group_changes {
+                // Recreate member list if the message comes from a MUA as these messages
+                // do _not_ set add/remove headers
+                if !mime_parser.has_chat_version() && apply_group_changes {
+                    true
+                } else {
+                    match mime_parser.get_header(HeaderDef::InReplyTo) {
+                        // If we don't know the referenced message, we missed some messages which could be add/delete
```

Comment:
When self is added, the recreate_membership_list flag is set and apply_groupchanges aswell. This will lead to a new better_msg in the header-handling code, so a stockstring should be created.

Reply:
But what if we did't receive a ChatGroupMemberAdded message about "self"? Afaiu, this PR is fixing problems when messages can be missed because of delivery problems. I don't get why "self" is a special case in ragards to handling of `InReplyTo` which references a missed message

Reply:
I moved it outwards where it belongs

Reply:
Anyway, consider the following scenario:
- We left the group
- Then we were re-added, but missed the addition message
- We receive a message with InReplyTo referencing a message we don't have (the message adding us to the group). But above there's the check:
```
if !chat::is_contact_in_chat(context, chat_id, ContactId::SELF).await? && !self_added {
```
So, we're not in the group and don't handle this scenario. Is it correct? Why is it special for SELF?

Reply:
IIRC at some point, there were lots of complaints that people couldn't leave a group because they were constantly re-added. Not sure if that's relevant here, didn't look at the code yet.

Reply:
For us to be unintentionally re-added two things should happen:
- Another user missed a message about our leave from the group and then sent a message (A) containing us in `To`
- We missed a message which A refers in `InReplyTo` and re-add ourselves to the group.

I think it's unlikely to happen and suggest not to protect from it. Moreover, in a way the PR suggests, because the scenario described above (when we miss our addition) is much more likely

Reply:
@Hocuri, i still don't get why we can't just use the algo you suggested initially in https://github.com/deltachat/deltachat-core-rust/issues/3782#issue-1466644721. For me it looks more simple and straightforward. Also it was discussed in the "DC Core Rust" chat, just search for "InReplyTo"

Reply:
Reading the thread here, tbh, I don't remember anymore what the problems discussed here were. But I think it's fine, IIRC we discussed pretty minor edge cases, time will tell if there are problems we should fix (and whether the new algo is an improvement at all). @iequidoo I hope you don't feel like we missed your concerns?

Reply:
The problem is that the current algo is not resillent to missed messages about addition of SELF. If it happens, the only way to fix the group is to remove the member and re-add them again. The algo you suggested in https://github.com/deltachat/deltachat-core-rust/issues/3782#issue-1466644721 fixes this inconsistency when we receive a message with orphaned InReplyTo. Also it looks simpler. And as for the problem with messages from a Delta Chat that restored an old backup, it will ignore them as well because they will reference a known message in InReplyTo.

I think it's not very important right now, but i'd prefer to try your "truly consistent" algo and if it doesn't bring new problems, why not use it?